File: ruby/lesson 1/pivot_index.rb
def pivot_index(nums)
    return -1 if nums.empty?

    sums = [nums[0]]
    (1..nums.length - 1).each { |i| sums[i] = sums[i - 1] + nums[i] }

    nums.each_with_index do |_, i|
        left = i.zero? ? 0 : sums[i - 1]
        right = sums[sums.length - 1] - sums[i]
        return i if left == right
    end

    -1
end

File: python/lesson 1/pivot_index.py
from typing import List


class Solution:
    def pivot_index(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return -1
        
        sums = [None] * len(nums)
        sums[0] = nums[0]
        for i in range(1, len(nums)):
            sums[i] = sums[i - 1] + nums[i]
            
        for i, _ in enumerate(nums):
            left =  0 if i == 0  else sums[i - 1]
            right = sums[len(sums) - 1] - sums[i]
            if left == right:
                return i
        
        return -1
